File: src/codes/Users_Manual/stm/foci_1.cpp
//~#include <ranges>
//~#include<autd3.hpp>
//~#include<autd3/link/nop.hpp>
//~using namespace std::ranges::views;
//~int main() {
//~auto autd =
//~autd3::ControllerBuilder({autd3::AUTD3(autd3::Vector3::Zero())}).open(autd3::link::Nop::builder());
const autd3::Vector3 center =
    autd.geometry().center() + autd3::Vector3(0, 0, 150);
const auto points_num = 200;
const auto radius = 30.0f;
auto stm = autd3::FociSTM<2>::from_freq(
    1.0f * autd3::Hz,
    iota(0) | take(points_num) | transform([&](auto i) {
      const auto theta = 2.0f * autd3::pi * static_cast<float>(i) /
                         static_cast<float>(points_num);
      autd3::Vector3 p = center + autd3::Vector3(radius * std::cos(theta),
                                                 radius * std::sin(theta), 0);
      return autd3::ControlPoints<2>(
                 std::array{autd3::ControlPoint(center + p).with_offset(0x00),
                            autd3::ControlPoint(center - p).with_offset(0x00)})
          .with_intensity(0xFF);
    }));
//~return 0; }

File: src/codes/Users_Manual/stm/foci_1.py
~import numpy as np
~from pyautd3 import AUTD3, Controller, FociSTM, Hz
from pyautd3 import ControlPoint, ControlPoints2
~from pyautd3.link.audit import Audit
~autd: Controller[Audit] = Controller.builder([AUTD3([0.0, 0.0, 0.0])]).open(
~    Audit.builder()
~)
center = autd.geometry.center + np.array([0.0, 0.0, 150.0])
point_num = 200
radius = 30.0
stm = FociSTM.from_freq(
    1.0 * Hz,
    (
        ControlPoints2(
            (
                ControlPoint(center + radius * np.array([np.cos(theta), np.sin(theta), 0])).with_offset(0x00),
                ControlPoint(center - radius * np.array([np.cos(theta), np.sin(theta), 0])).with_offset(0x00),
            ),
        ).with_intensity(0xFF)
        for theta in (2.0 * np.pi * i / point_num for i in range(point_num))
    ),
)